plugins/InspectionGadgets/InspectionGadgetsAnalysis/src/com/siyeh/ig/bugs/BaseEqualsVisitor.java

```diff
@@ -53,19 +67,156 @@ public void visitMethodReferenceExpression(PsiMethodReferenceExpression expressi
     }
     else if (STATIC_EQUALS.test(expression)) {
-      expression1 = PsiUtil.skipParenthesizedExprDown(arguments[0]);
-      expression2 = PsiUtil.skipParenthesizedExprDown(arguments[1]);
+      PsiExpression expression1 = PsiUtil.skipParenthesizedExprDown(arguments[0]);
+      PsiExpression expression2 = PsiUtil.skipParenthesizedExprDown(arguments[1]);
+      checkTypes(expression, expression1, expression2);
     }
-    else {
-      return;
+    else if (PREDICATE_IS_EQUAL.test(expression)) {
+      PsiExpression expression1 = PsiUtil.skipParenthesizedExprDown(arguments[0]);
+      PsiType psiType1 = getType(expression1);
+      PsiType psiType2 = resolveIsEqualPredicateType(expression);
+      if (psiType1 == null || psiType2 == null) {
+        return;
+      }
+      checkTypes(expression.getMethodExpression(), psiType1, psiType2);
+    }
+  }
+
+  @Nullable
+  private static PsiType resolveIsEqualPredicateType(PsiMethodCallExpression expression) {
+    PsiExpression upperPredicate = expression;
+    PsiElement parent = upperPredicate.getParent();
+    int max = 100;
+    int currentLevel = 0;
+    while (currentLevel <= max) {
+      currentLevel++;
+      if (parent instanceof PsiParenthesizedExpression || parent instanceof PsiReferenceExpression) {
+        parent = parent.getParent();
+      }
+      else if (parent instanceof PsiExpressionList && isPredicateSourceForIsEqual(parent.getParent())) {
+        upperPredicate = (PsiExpression)parent.getParent();
+        parent = upperPredicate.getParent();
+      }
+      else if (isPredicateSourceForIsEqual(parent)) {
+        upperPredicate = (PsiExpression)parent;
+        parent = upperPredicate.getParent();
+      }
+      else {
+        break;
+      }
+    }
+
+    return findPsiTypeForPredicate(upperPredicate, parent);
+  }
+
+  @Nullable
+  private static PsiType findPsiTypeForPredicate(PsiExpression upperPredicate, PsiElement parent) {
+    PsiType returnParameter = findPredicateParameter(upperPredicate);
+    if (returnParameter != null) {
+      return returnParameter;
     }
+
+    if (parent instanceof PsiMethodCallExpression) {
+      final PsiMethodCallExpression callExpresion = (PsiMethodCallExpression)parent;
+      if (PREDICATE_TEST.test(callExpresion)) {
+        final PsiExpression[] argumentExpressions = callExpresion.getArgumentList().getExpressions();
+        if (argumentExpressions.length != 1) {
+          return null;
+        }
+        PsiExpression argument = PsiUtil.skipParenthesizedExprDown(argumentExpressions[0]);
+        return getType(argument);
+      }
+    }
+    else if (parent instanceof PsiExpressionList) {
+      return findPredicateTypeFromExpressionList(upperPredicate, (PsiExpressionList)parent);
+    }
+    return null;
+  }
+
+  @Nullable
+  private static PsiType findPredicateTypeFromExpressionList(PsiExpression upperPredicate, PsiExpressionList expressionList) {
+    Integer index = findParameterIndex(expressionList, upperPredicate);
```

Comment:
I think, `MethodCallUtils.getParameterForArgument` with subsequent `PsiUtil.substituteTypeParameter` would do most of the work you need. There are also more generic utilities like `ExpectedTypeUtils.findExpectedType`.

Reply:
Rewrite with `ExpectedTypeUtils.findExpectedType` - thank you. This method is awesome!